git/archive.go

```diff
@@ -0,0 +1,260 @@
+		if t, err := CommitID(sha).GetDate(c); err == nil {
+			mtime = t
+		}
+	}
+
+	// If gzip compression is enabled
+	if tgz {
+		gw := gzip.NewWriter(fileOutput)
+		fileOutput = gw
+		defer gw.Close()
+	}
+
+	// Create the tar writer
+	tw := tar.NewWriter(fileOutput)
+	defer tw.Close()
+
+	// Write the pax header
+	hdr := &tar.Header{
+		Typeflag: tar.TypeXGlobalHeader,
+		Name:     "pax_global_header",
+		PAXRecords: map[string]string{
+			"comment": sha.String(),
+		},
+	}
+
+	if err := tw.WriteHeader(hdr); err != nil {
+		return err
+	}
+
+	for _, e := range entries {
+		o, err := c.GetObject(e.Sha1)
+		if err != nil {
+			return err
+		}
+		if obj, ok := o.(GitBlobObject); ok {
+			hdr := &tar.Header{}
+			hdr.Name = opts.BasePrefix + e.PathName.String()
+			hdr.Size = int64(obj.GetSize())
+			hdr.ModTime = mtime
+
+			// TODO: Mask the mode. by default the mask is 0002 (turn off write bit)
+			// but can be changed using tar.umask config.
+			switch e.Mode {
+			case ModeBlob:
+				hdr.Mode = 0644
+			case ModeExec:
+				hdr.Mode = 0755
+			default:
+				hdr.Mode = 0644
+			}
+
+			if err := tw.WriteHeader(hdr); err != nil {
+				return err
+			}
+			if _, err := tw.Write(o.GetContent()); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func createZipArchive(c *Client, opts ArchiveOptions, sha Sha1, entries []*IndexEntry) error {
+	fileOutput := os.Stdout
+
+	// If the output file is not empty use it instead of stdout
+	if opts.OutputFile != "" {
+		if file, err := os.Create(opts.OutputFile); err == nil {
+			fileOutput = file
+			defer file.Close()
+		} else {
+			return err
+		}
+	}
+
+	mtime := time.Now()
+
+	// If the sha is a tree we set the file modification time to the current time
+	// otherwise we must use the commit time.
+	if sha.Type(c) != "tree" {
+		if t, err := CommitID(sha).GetDate(c); err == nil {
+			mtime = t
+		}
+	}
+
+	zw := zip.NewWriter(fileOutput)
+	defer zw.Close()
+
+	// set the compression level
+	zw.RegisterCompressor(zip.Deflate, func(out io.Writer) (io.WriteCloser, error) {
+		return flate.NewWriter(out, opts.CompressionLevel)
+	})
+
+	// Set the commit sha as the zip comment.
+	zw.SetComment(sha.String())
+
+	for _, e := range entries {
+		o, err := c.GetObject(e.Sha1)
+		if err != nil {
+			return err
+		}
+		if obj, ok := o.(GitBlobObject); ok {
+			hdr := &zip.FileHeader{
+				Name:     opts.BasePrefix + e.PathName.String(),
+				Modified: mtime,
+				Method:   zip.Deflate,
+			}
+			f, err := zw.CreateHeader(hdr)
+
+			if err != nil {
+				return nil
+			}
+
+			if _, err := f.Write(obj.GetContent()); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// Return the list of supported archive file format
+func ArchiveFormatList() []string {
+	formats := make([]string, 0, len(supportedArchiveFormats))
+	for key := range supportedArchiveFormats {
+		formats = append(formats, key)
+	}
+	return formats
+}
+
+func Archive(c *Client, opts ArchiveOptions, arg string) error {
```

Comment:
If you make arg a Treeish instead of string and move the conversion into package string, you could simplify this code. (You could also implement paths by just adding a []File argument here that you pass along the LsTree..)